### PR TITLE
DEVPROD-5753: Implement and apply requireHostAccess GQL directive

### PR DESCRIPTION
**Pull request by SupaJoon**
DEVPROD-5753

### Description
The actual added code to review is less than 540 lines. 
These code changes implement and apply the `@requireHostAccess` directive to these mutations: 

- updateSpawnHostStatus
- editSpawnHost
- reprovisionToNew
- restartJasper
- updateHostStatus

and these queries:

- host
- hostEvents

I also applied the directive permission logic to the `hosts` query resolver instead of the schema since there is are no host IDs available before the host db query takes place.

These code changes combine the permission for editing spawn hosts and non-spawn hosts. Currently in Evergreen, the difference between the two implementations is that spawn hosts permission is granted if the host startedBy field is the same as the requesting user and the user has the host edit permission ([code](https://github.com/evergreen-ci/evergreen/blob/bd7a267564e978df0e6a545f8e0a69b345a829a7/model/host/host.go#L795)) whereas modifying a non-spawn host only checks the edit permission ([code](https://github.com/evergreen-ci/evergreen/blob/bd7a267564e978df0e6a545f8e0a69b345a829a7/api/host.go#L77)). I confirmed in staging that I am able to modify my spawn host with the current code changes and I could use edge case clarification from an App engineer. 

e2e test is fixed in this UI PR: https://github.com/evergreen-ci/ui/pull/410

### CODE REVIEW

File: graphql/query_resolver.go [hunk 1]
```diff
@@ -338,14 +338,23 @@ func (r *queryResolver) Hosts(ctx context.Context, hostID *string, distroID *str
...
-	apiHosts := []*restModel.APIHost{}
+	usr := mustHaveUser(ctx)
+	forbiddenHosts := []string{}
+	for _, h := range hosts {
+		if !userHasHostPermission(usr, h.Distro.Id, evergreen.HostsView.Value) {
+			forbiddenHosts = append(forbiddenHosts, h.Id)
+		}
+	}
+	if len(forbiddenHosts) > 0 {
+		return nil, Forbidden.Send(ctx, fmt.Sprintf("user '%s' does not have permission to access one or more hosts.", usr.Username()))
```
Comment (minnakt) [inline graphql/query_resolver.go:349]: nit. formatting (no period)
```suggestion
		return nil, Forbidden.Send(ctx, fmt.Sprintf("user '%s' does not have permission to access one or more hosts", usr.Username()))
```

Comment (minnakt) [inline graphql/query_resolver.go:349]: since we don't really do anything with the `forbiddenHosts`, I wonder if it makes sense to do this check in the same loop where we're building the APIHosts below (and just return immediately if you encounter no permissions)

File: graphql/resolver.go [hunk 2]
```diff
@@ -35,6 +36,46 @@ func New(apiURL string) Config {
...
+		hostIdsInterface, hasHostIds := args["hostIds"].([]interface{})
+		hostIds := []string{}
+		if hasHostIds {
+			for _, v := range hostIdsInterface {
+				hostIds = append(hostIds, v.(string))
+			}
+		}
+		if !hasHostId && !hasHostIds {
+			return nil, ResourceNotFound.Send(ctx, "host not specified")
+		}
```
Comment (minnakt) [inline graphql/resolver.go:54]: I think it makes sense to switch the order, since we might not end up using `hostIds` if we hit this if statement
```suggestion
		if !hasHostId && !hasHostIds {
			return nil, ResourceNotFound.Send(ctx, "host not specified")
		}
		
		hostIds := []string{}
		if hasHostIds {
			for _, v := range hostIdsInterface {
				hostIds = append(hostIds, v.(string))
			}
		}
```

File: graphql/resolver.go [hunk 3]
```diff
@@ -35,6 +36,46 @@ func New(apiURL string) Config {
 			sc: dbConnector,
 		},
 	}
+	c.Directives.RequireHostAccess = func(ctx context.Context, obj interface{}, next graphql.Resolver, access HostAccessLevel) (interface{}, error) {
+		args, isStringMap := obj.(map[string]interface{})
+		if !isStringMap {
+			return nil, ResourceNotFound.Send(ctx, "host not specified")
+		}
+		hostId, hasHostId := args["hostId"].(string)
+		hostIdsInterface, hasHostIds := args["hostIds"].([]interface{})
```
Comment (minnakt) [inline graphql/resolver.go:45]: does it make sense to do this so you can skip the extra loop for `hostIds` 👀
```suggestion
		hostIds, hasHostIds := args["hostIds"].([]string)
```

File: graphql/directive_test.go [hunk 4]
```diff
@@ -218,7 +234,87 @@ func setupPermissions(t *testing.T) {
 	}
 	require.NoError(t, roleManager.UpdateRole(logViewRole))
 }
-
+func TestRequireHostAccess(t *testing.T) {
```
Comment (hadjri) [inline graphql/directive_test.go:237]: The references to `.Equal` or `.EqualError` in this file should use `assert` rather than `require` for codebase style consistency.

File: graphql/query_resolver.go (hunk 1 above)
Comment (hadjri) [reply graphql/query_resolver.go:349]: You can either do that or include a formatted list of all the forbidden hosts in the error message [here](https://github.com/evergreen-ci/evergreen/pull/8337/files#diff-18582cc382c2280a0446b8fa448fd4ca997753d6ebfe6042a981701b312164c6R349)

File: graphql/resolver.go [hunk 5]
```diff
@@ -35,6 +36,46 @@ func New(apiURL string) Config {
...
+		hostIds := []string{}
+		if hasHostIds {
+			for _, v := range hostIdsInterface {
+				hostIds = append(hostIds, v.(string))
+			}
+		}
+		if !hasHostId && !hasHostIds {
+			return nil, ResourceNotFound.Send(ctx, "host not specified")
+		}
+		hostIdsToCheck := []string{}
```
Comment (hadjri) [inline graphql/resolver.go:55]: nit: initialize as `[]string{hostId}` and remove L56

File: graphql/resolver.go [hunk 6]
```diff
@@ -35,6 +36,46 @@ func New(apiURL string) Config {
...
+		hostIdsToCheck = append(hostIdsToCheck, hostId)
+		hostIdsToCheck = append(hostIdsToCheck, hostIds...)
+		var requiredLevel int
+		if access == HostAccessLevelEdit {
+			requiredLevel = evergreen.HostsEdit.Value
+		} else if access == HostAccessLevelView {
```
Comment (hadjri) [inline graphql/resolver.go:61]: is access not being specified a valid case? If not I would use `else` rather than `else if`

File: graphql/resolver.go [hunk 7]
```diff
@@ -35,6 +36,46 @@ func New(apiURL string) Config {
...
+		user := mustHaveUser(ctx)
+		hostsToCheck, err := host.Find(ctx, host.ByIds(hostIdsToCheck))
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, "Error getting hosts")
+		}
+		if len(hostsToCheck) == 0 {
+			return nil, ResourceNotFound.Send(ctx, "No matching hosts found")
+		}
+		for _, h := range hostsToCheck {
+			if !userHasHostPermission(user, h.Distro.Id, requiredLevel) {
```
Comment (hadjri) [inline graphql/resolver.go:73]: This should be consistent with the logic [here](https://github.com/evergreen-ci/evergreen/pull/8337/files#diff-18582cc382c2280a0446b8fa448fd4ca997753d6ebfe6042a981701b312164c6R344) (as in whether we keep track of a list of forbidden hosts or not)

**Review by hadjri — CHANGES_REQUESTED** [review]

File: graphql/query_resolver.go (hunk 1 above)
Comment (SupaJoon) [reply graphql/query_resolver.go]: I originally considered returning a list of hosts that are inaccessible but I think that is insecure since the query parameters match the hostIds in the error message.

**Comment by SupaJoon** [comment]
evergreen retry

File: graphql/resolver.go [hunk 8]
```diff
@@ -35,6 +36,54 @@ func New(apiURL string) Config {
...
+		if access == HostAccessLevelEdit {
+			requiredLevel = evergreen.HostsEdit.Value
+		} else {
+			requiredLevel = evergreen.HostsView.Value
+		}
+		user := mustHaveUser(ctx)
+		hostsToCheck, err := host.Find(ctx, host.ByIds(hostIdsToCheck))
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, "Error getting hosts")
+		}
```
Comment (minnakt) [inline graphql/resolver.go:68]: should we include the `err` message here?

File: graphql/resolver.go [hunk 9]
```diff
@@ -35,6 +36,54 @@ func New(apiURL string) Config {
...
+			if !userHasHostPermission(user, h.Distro.Id, requiredLevel) {
+				forbiddenHosts = append(forbiddenHosts, h.Id)
+			}
+		}
+		if len(forbiddenHosts) == 1 {
+			return nil, Forbidden.Send(ctx, fmt.Sprintf("user '%s' does not have permission to access host '%v'", user.Username(), forbiddenHosts[0]))
+		} else if len(forbiddenHosts) > 1 {
+			hostsString := strings.Join(forbiddenHosts, ", ")
+			return nil, Forbidden.Send(ctx, fmt.Sprintf("user '%s' does not have permission to access hosts: '%v'", user.Username(), hostsString))
+		}
```
Comment (minnakt) [inline graphql/resolver.go:83]: could this be consolidated? I also thing usink `%s` might be slightly better than using `%v` since it's a string
```suggestion
		if len(forbiddenHosts) > 0 {
			hostsString := strings.Join(forbiddenHosts, ", ")
			return nil, Forbidden.Send(ctx, fmt.Sprintf("user '%s' does not have permission to access host(s): '%s'", user.Username(), hostsString))
		}
```

**Review by hadjri — APPROVED** [review]

File: graphql/resolver.go [hunk 10]
```diff
@@ -35,6 +36,54 @@ func New(apiURL string) Config {
...
+
+		hostIds := []string{}
+		if hasHostIds {
+			for _, v := range hostIdsInterface {
+				hostIds = append(hostIds, v.(string))
+			}
+		}
+		hostIdsToCheck := []string{hostId}
+		hostIdsToCheck = append(hostIdsToCheck, hostIds...)
```
Comment (minnakt) [inline graphql/resolver.go:57]: Could the same effect be achieved by:
```suggestion
		hostIdsToCheck := []string{hostId}
		if hasHostIds {
			for _, v := range hostIdsInterface {
				hostIdsToCheck = append(hostIdsToCheck, v.(string))
			}
		}
```

**Review by minnakt — APPROVED** [review]

File: graphql/resolver.go (hunk 9 above)
Comment (SupaJoon) [reply graphql/resolver.go:81]: This could be consolidated but I like how the copy is more specific in the original code. I updated %v to %s.
